### PR TITLE
Support `.mjs` plugins/presets and async factories

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | 
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

This is the first step needed to migrate to native ECMAScript modules: we must support loading `.mjs` plugins and presets :stuck_out_tongue:

- The first commit is mostly needed for third-party presets: many of them directly use `require()` inside the preset factory, and this makes it possible to use `await import()` instead.
- The second commit enables loading `.mjs` plugins/presets. It's mostly functions being converted to generators to work with `gensync`.

While preparing this PR, I noticed that we currently support this kind of presets:
```js
exports.__esModule = true;
exports.plugins = [foo];
```
which is the result of transpiling
```js
export const plugins = [foo];
```

However, we throw an error for
```js
exports.__esModule = true
exports.default = { plugins: [foo] }
```
because presets must export a function, and not a plain object:
```js
exports.__esModule = true
exports.default = function () {
  return { plugins: [foo] };
};
```

Since presets must export a function (so that they can properly configure caching), I didn't implement support for the named exports in `.mjs` files.

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/12266"></a>

### CODE REVIEW

File: packages/babel-core/test/async.js [hunk 1]
```diff
@@ -7,6 +9,21 @@ const nodeGte8 = (...args) => {
...
 };
 
+async function compileInSpawedProcess(async) {
+  // !!!! hack is coming !!!!
+  // todo(Babel 8): remove this section when https://github.com/facebook/jest/issues/9430 is resolved
+  // We don't check process.versions.node here, it will fail if node does not support esm
+  // please publish Babel on a modernized node :)
+  const { stdout, stderr } = await util.promisify(cp.execFile)(
+    require.resolve(`./fixtures/babel-compile-${async ? "async" : "sync"}.mjs`),
+    { env: process.env },
```
Comment (JLHwung) [inline packages/babel-core/test/async.js:19]: nit: By default `env` is passed with `process.env`. https://nodejs.org/api/child_process.html#child_process_child_process_execfile_file_args_options_callback

Comment (nicolo-ribaudo) [reply packages/babel-core/test/async.js:19]: Actually, I need someone with macOS to check if I need to do https://github.com/babel/babel/blame/31396b286daadee3c25c4891c8dd31a92d7f0604/packages/babel-core/test/config-chain.js#L63 for this test.

Comment (JLHwung) [reply packages/babel-core/test/async.js:19]: `make prepublish` is passing as of dfa75fa65f on macOS.

**Review by hzoo — APPROVED** [review]

**Review by JLHwung — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
I'm merging this PR to the `feat-7.13.0/babel-core-features` branch, because there are three different PRs that modify `@babel/core` that conflict with each other (this one, `assumptions` and `targets`)

**Pull request merged**